### PR TITLE
handle ioctl errno

### PR DESCRIPTION
**Pull request by gnode1**
Turns out modern KVM `ioctl (KVM_RUN)` _sometimes_ returns -errno and keeps proper errno untouched. According to `man ioctl()` it should return -1 and set `errno` in case of an error. However I observe that ioctl returning -14 (EFAULT) and errno is left untouched.

### CODE REVIEW

**Review by johnmuth81 — APPROVED** [review]
KM is public, so I guess I still see it.

LGTM ;)

But I don't have write access, so my approval means nothing :(

File: km/km_vcpu_run.c
Comment (paulpopelka) [inline km/km_vcpu_run.c:513]: This does need a comment to the effect that we are compensating for some kind of bug in ioctl()'s ability to set errno properly.

Comment (sv641) [reply km/km_vcpu_run.c:513]: errno most probably has changed between line 499 - 513(depending on what options we run with). we need to save it between lines 499 and 500.

**Review by paulpopelka — APPROVED** [review]
Just one minor comment.

And, github says this is being committed with my github id.  Is there a workspace on my machine that contains this?

File: km/km_vcpu_run.c
Comment (paulpopelka) [reply km/km_vcpu_run.c:513]: Then this needs to be fixed differently.
Save errno after ioctl() returns and use that value to restore errno.

Plus ioctl() should not return -errno on failure but that is apparently what we see.

Comment (sv641) [reply km/km_vcpu_run.c:513]: `int
__ioctl (int fd, unsigned long int request, ...)
{
  va_list args;
  va_start (args, request);
  void *arg = va_arg (args, void *);
  va_end (args);

  int r;
  if (!__ioctl_arch (&r, fd, request, arg))
    {
      r = INTERNAL_SYSCALL_CALL (ioctl, fd, request, arg);
      if (__glibc_unlikely (INTERNAL_SYSCALL_ERROR_P (r)))
	{
	  __set_errno (-r);
	  return -1;
	}
    }
  return r;
}`

Comment (paulpopelka) [inline km/km_vcpu_run.c:513]: I think the proper fix is to save errno in a local variable  immediately after ioctl( KVM_RUN ) returns.
And then when we discover ioctl() has failed, restore errno so that km_trace() will display it properly and the switch statement below this will be operating on a valid value.

**Review by paulpopelka — COMMENTED** [review]
I think you brought this fix over from km-gpu because it was thought to be some sort of problem with how ioctl() worked.
We have since found that the ioctl() hook we are using in km-gpu was causing the problem we are trying to fix here.
Since we don't hook ioctl() in this repository this fix does not apply here.

But, we do need to save errno since there are several lines of code between the call to ioctl( KVM_RUN ) and the point where we want to look at errno.  And some of those intervening lines can modify errno.

File: km/km_vcpu_run.c
Comment (gnode1) [reply km/km_vcpu_run.c:513]: you right, will do

File: tests/km_core_tests.bats
Comment (paulpopelka) [inline tests/km_core_tests.bats:1681]: Shouldn't this be " wait $pid" ?

File: km/km_vcpu_run.c
Comment (paulpopelka) [inline km/km_vcpu_run.c:520]: I think you should restore errno.
The invocation of km_err() in the default cause wants to use errno.

File: km/km_vcpu_run.c
Comment (paulpopelka) [inline km/km_vcpu_run.c:498]: Is this really needed?

Comment (gnode1) [reply km/km_vcpu_run.c:498]: ioctl doesn't set it if there is no error. So the code below would act on old value

**Pull request merged**
